ush/python/pygw/src/pygw/executable.py

```diff
@@ -0,0 +1,315 @@
+#
+import os
+import shlex
+import subprocess
+import sys
+from typing import Any, Optional
+
+__all__ = ["Executable", "which", "CommandNotFoundError"]
+
+
+class Executable:
+    """
+    Class representing a program that can be run on the command line.
+
+    Example:
+    --------
+
+    >>> from pygw.executable import Executable
+    >>> cmd = Executable('srun')  # Lets say we need to run command e.g. "srun"
+    >>> cmd.add_default_arg('my_exec.x')  # Lets say we need to run the executable "my_exec.x"
+    >>> cmd.add_default_arg('my_arg.yaml')  # Lets say we need to pass an argument to this executable e.g. "my_arg.yaml"
+    >>> cmd.add_default_env('OMP_NUM_THREADS', 4)  # Lets say we want to run w/ 4 threads in the environment
+    >>> cmd(output='stdout', error='stderr')  # Run the command and capture the stdout and stderr in files named similarly.
+
+    `cmd` line above will translate to:
+
+    $ export OMP_NUM_THREADS=4
+    $ srun my_exec.x my_arg.yaml 1>&stdout 2>&stderr
+
+    References
+    ----------
+    .. [1] "spack.util.executable.py", https://github.com/spack/spack/blob/develop/lib/spack/spack/util/executable.py
+    """
+
+    def __init__(self, name: str):
+        """
+        Construct an executable object.
+
+        Parameters
+        ----------
+        name : str
+            name of the executable to run
+        """
+        self.exe = shlex.split(str(name))
+        self.default_env = {}
+        self.returncode = None
+
+        if not self.exe:
+            raise ProcessError(f"Cannot construct executable for '{name}'")
+
+    def add_default_arg(self, arg: Any) -> None:
+        """Add a default argument to the command."""
+        self.exe.append(arg)
+
+    def add_default_env(self, key: str, value: Any) -> None:
```

Comment:
I think `default_env` makes sense because the `__call__` method takes optional keyword arguments `env` to over-ride or add to the default environment.  If `env` is passed, then the executable is run with that environment (`env`) + `default_env`.  If `env` is `None`, the executable is run with the `os.environ` + `default_env`

Similarly, one can add default args to the `exe` as well as pass command specific args via the `__call__` magic function.